### PR TITLE
Add gradients and borders to FormSpec boxes

### PR DESCRIPTION
**Pull request by v-rob**
This PR adds gradients and borders to FormSpec boxes using the styles `colors`, `bordercolors`, and `borderstyles`.  If a color is specified in a `box` definition, that color will override the `boxoptions` color for that box.

Each option can either take a single value that applies to all four sides, or four comma-separated values applying to each side in clockwise order.  For the box color, the first value is the top-left corner color.  The other two options start with the top border.

Unlike normal boxes, the colors in `style_type` will not be semitransparent by default.  This is because it is more specialized, and therefore should have to be more explicit.

Borders are specified in pixels since borders are often very thin; if a normal coordinate is needed, use multiple boxes.  Each border can be either positive (it extends out of the box area), negative (it takes up space inside the box area), or zero (no border at all).  Any combination for any border will work.  The borders and the box will never overlap, nor will the borders overlap each other, even with a negative value.

## Screenshots

image

Very fancy gradients and borders are possible, as you can see.

gradients

For the two boxes with the full rainbow, six boxes had to be used.  But as you can see, they merge seamlessly.

<details>
<summary><b>Code:</b></summary>

```lua
"size[10,10]"..
"real_coordinates[true]"..

"button[0,0;1,1;;In (-10)]"..
"style_type[box;colors=red,yellow,green,blue;bordercolors=#FF000055,#0000FF55,#00FF0055,#FF00FF55;borderwidths=-10]"..
"box[1,1;3,3;]"..

"button[5,0;1,1;;Out (+10)]"..
"style_type[box;bordercolors=#FF000055,#0000FF55;borderwidths=10]"..
"box[6,1;3,3;]"..

"button[1,5;3,1;;Uneven Border]"..
"style_type[box;colors=#ffffff8c;bordercolors=#ffffff8c;borderwidths=5,10,15,20]"..
"box[1,5;3,1;]"..

"style_type[box;colors=;borderwidths=2]"..
"box[5,5;4,4;]"..
"label[5.15,4.85;Box Area]"..
"field[5.25,5.5;3.5,0.5;name;Name:;]"..
"field[5.25,6.5;3.5,0.5;date;Date:;]"..
"field[5.25,7.5;3.5,0.5;whatever;Stuff:;]"..
"button[6.75,8.25;2,0.5;submit;Submit]"..

"style_type[box;colors=blue,blue,#0000FF00,#0000FF00;borderwidths=0]"..
"box[1,7;3,2;]"
```

```lua
"size[4.5,7]"..
"real_coordinates[true]"..
"style_type[box;colors=lightseagreen,lightseagreen,#0000008c,#0000008c]"..
"box[0,-1;4.5,1;]"..
"style_type[box;colors=#0000008c,#0000008c,lightseagreen,lightseagreen]"..
"box[0,7;4.5,1;]"..
"style_type[box;bordercolors=#20B2AA55;borderwidths=1,0,-1,0]"..
"box[0,3.5;4.5,0;]"..

"container[0.25,0.25]"..
"style_type[box;colors=black,#FF0000,white,black;bordercolors=black;borderwidths=1]"..
"box[0,0;3,3;]"..
"box[3.25,0;0.25,3;#00000000]"..
"style_type[box;colors=#FF0000,#FF0000,#FF000000,#FF000000]"..
"box[3.75,0;0.25,3;]"..
"style_type[box;colors=#FF0000,#FF0000,#FFFF00,#FFFF00;borderwidths=]"..
"box[3.25,0;0.25,0.5;]"..
"style_type[box;colors=#FFFF00,#FFFF00,#00FF00,#00FF00]"..
"box[3.25,0.5;0.25,0.5;]"..
"style_type[box;colors=#00FF00,#00FF00,#00FFFF,#00FFFF]"..
"box[3.25,1;0.25,0.5;]"..
"style_type[box;colors=#00FFFF,#00FFFF,#0000FF,#0000FF]"..
"box[3.25,1.5;0.25,0.5;]"..
"style_type[box;colors=#0000FF,#0000FF,#FF00FF,#FF00FF]"..
"box[3.25,2;0.25,0.5;]"..
"style_type[box;colors=#FF00FF,#FF00FF,#FF0000,#FF0000]"..
"box[3.25,2.5;0.25,0.5;]"..
"container_end[]"..

"container[0.25,3.75]"..
"style_type[box;colors=;bordercolors=black;borderwidths=1]"..
"box[0,0;3,3;]"..
"style_type[box;colors=#FF0000,#FFFF00,#808080,#808080;borderwidths=]"..
"box[0,0;0.5,3;]"..
"style_type[box;colors=#FFFF00,#00FF00,#808080,#808080]"..
"box[0.5,0;0.5,3;]"..
"style_type[box;colors=#00FF00,#00FFFF,#808080,#808080]"..
"box[1,0;0.5,3;]"..
"style_type[box;colors=#00FFFF,#0000FF,#808080,#808080]"..
"box[1.5,0;0.5,3;]"..
"style_type[box;colors=#0000FF,#FF00FF,#808080,#808080]"..
"box[2,0;0.5,3;]"..
"style_type[box;colors=#FF00FF,#FF0000,#808080,#808080]"..
"box[2.5,0;0.5,3;]"..
"style_type[box;colors=white,white,black,black;bordercolors=black;borderwidths=1]"..
"box[3.25,0;0.25,3;]"..
"style_type[box;colors=#FF0000,#FF0000,#FF000000,#FF000000]"..
"box[3.75,0;0.25,3;]"..
"container_end[]"
```
</details>

## To do

This PR is Ready for Review.

- [x] Integrate with `style`
- [x] Add documentation

### CODE REVIEW

**Comment by paramat** [comment]
@v-rob we might close this if it gets to the point of no progress for 6 months.

**Comment by v-rob** [comment]
Well, here's your progress :)

I've gotten it mostly correct, since it compiles correctly, and the Minetest window pops up momentarily, but then I get `LLVM Error: out of memory`, so I'm working on that now.

**Comment by v-rob** [comment]
This PR is now ready for review.

File: src/gui/guiBox.cpp [hunk 1]
```diff
@@ -31,8 +34,82 @@ void GUIBox::draw()
...
+
+	for (unsigned int i = 0; i <= 3; i++) {
+		if (m_borderwidths[i] > 0)
+			outer_bd[i] = m_borderwidths[i];
+		else
+			inner_bd[i] = m_borderwidths[i];
+	}
+
+	v2s32 pos = AbsoluteRect.UpperLeftCorner;
+	v2s32 geom = AbsoluteRect.LowerRightCorner;
```
Comment (SmallJoker) [inline src/gui/guiBox.cpp]: These should rather be called `upperleft` and `lowerright` to not overcomplicate the variable naming.

File: src/gui/guiBox.cpp [hunk 2]
```diff
@@ -31,8 +34,82 @@ void GUIBox::draw()
...
+
+	v2s32 bd_pos = {
```
Comment (SmallJoker) [inline src/gui/guiBox.cpp]: `topleft_outer` would be way easier to understand here.

File: src/gui/guiBox.cpp [hunk 3]
```diff
@@ -31,8 +34,82 @@ void GUIBox::draw()
...
+		pos.X - outer_bd[3],
+		pos.Y - outer_bd[0]
+	};
+
+	v2s32 rect_pos = {
```
Comment (SmallJoker) [inline src/gui/guiBox.cpp]: `topleft_inner` or similar. Same for the variables below.

File: src/gui/guiBox.cpp [hunk 4]
```diff
@@ -31,8 +34,82 @@ void GUIBox::draw()
...
+	};
+
+	core::rect<s32> main_rect(
+		rect_pos.X,
+		rect_pos.Y,
+		rect_geom.X,
+		rect_geom.Y
+	);
+
+	std::vector<core::rect<s32>> border_rects(4);
```
Comment (SmallJoker) [inline src/gui/guiBox.cpp]: Linked list abuse.
`core::rect<s32> border_rects[4];` is good enough.

File: src/gui/guiFormSpecMenu.cpp [hunk 5]
```diff
@@ -2084,31 +2084,89 @@ void GUIFormSpecMenu::parseBox(parserData* data, const std::string &element)
...
+
 		video::SColor tmp_color;
+		std::vector<video::SColor> colors;
+		std::vector<video::SColor> bordercolors;
+		std::vector<s32> borderwidths;
+
+		// Colors
+		if (parseColorString(parts[2], tmp_color, true, 0x8C))
+			colors = {tmp_color, tmp_color, tmp_color, tmp_color};
+		else {
```
Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.cpp]: https://www.kernel.org/doc/html/latest/process/coding-style.html#placing-braces-and-spaces

Please put brackets around the first code block as well. Same for L2117, L2124, ...

File: src/gui/guiBox.cpp (hunk 4 above)
Comment (v-rob) [reply src/gui/guiBox.cpp]: C++ ruins my brain.  I don't know why I don't do these things.

Comment (rubenwardy) [reply src/gui/guiBox.cpp]: Vectors don't use linked list, and a std::array is usually preferable to a raw array

**Comment by v-rob** [comment]
Rebased, and I added a new feature where an array of two applies to the top/bottom and the left/right, like two parameters to a CSS border.  I also fixed a bug where the borders went top, bottom, right, left instead of top, right, bottom, left in clockwise order like they should.  The testing code has been fixed as well and also tests the two-array syntax.

As for why the box borders appeared that way with my testing code:  I had this weird bug where `style` and `style_type` all values would be reset every time a new element was specified even if I didn't change it.  My testing code used this incorrect behaviour.  This happened in multiple of my PRs, and it only just fixed itself recently.  It must have had something to do with the way I specifically was compiling Minetest.  Or a bug in the virtual machine, which I've been having a few problems with.  I don't know, but it's fixed now.

**Comment by v-rob** [comment]
I've moved the code around so that the array handling has been moved to `StyleSpec` (`getColorArray` and `getIntegerArray`) where it more properly belongs.  I also changed all `std::vectors` to `std::arrays` where possible.

File: doc/lua_api.txt [hunk 6]
```diff
@@ -2634,6 +2633,19 @@ Some types may inherit styles from parent types.
 * tabheader
     * noclip - boolean, set to true to allow the element to exceed formspec bounds.
     * textcolor - color. Default white.
+* box
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Please keep the alphabetic sorting.

File: doc/lua_api.txt [hunk 7]
```diff
@@ -2634,6 +2633,19 @@ Some types may inherit styles from parent types.
...
+    * noclip - boolean, set to true to allow the element to exceed formspec bounds.
+        * Defaults to false in formspec_version version 3 or higher
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Please also document this in `networkprocol.h`.

File: doc/lua_api.txt [hunk 8]
```diff
@@ -2634,6 +2633,19 @@ Some types may inherit styles from parent types.
...
+    * **Note**: `colors`, `bordercolors`, and `borderwidths` can either be a single value
+      like `red`, an array of two like `red,yellow` (applies to the top/bottom and
+      right/left or top-left/bottom-right and top-right/bottom-left), or an array of four
+      like `red,yellow,green,blue` (applies to the top or top-left and rotates around
+      clockwise). This works the same as CSS borders.
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Suggestion:
```
    * **Note**: `colors`, `bordercolors`, and `borderwidths` accept multiple input types:
        * Single value (ex. `#FF0`): All borders
        * Two values (ex. `red,#FFAAFF`): Top/border and left/right border
        * Four values: Top border and rotates around clockwise
        * This is comparable to CSS borders.
```

Q: How do you specify whether it starts from the top, or does gradients starting from top-left?

Comment (v-rob) [reply doc/lua_api.txt]: Gradients do start at the top-left.  It's in the current documentation.

**Comment by v-rob** [comment]
Ok, the pink was coming from a box behind the other boxes, so the pink color isn't the problem here, and I've just fixed that.  But the boxes do seem to have single pixel spaces between them.  If you place a button as wide as the form, it is cut off by one pixel, so the form width is too short.  I'm guessing this has to do with `imgsize` rounding errors because it affects clipping and because they go away at different screen sizes, so this might not be this PR's fault.

File: doc/lua_api.txt (hunk 6 above)
Comment (v-rob) [reply doc/lua_api.txt]: The sorting isn't alphabetical as it is.  Does `s` come after `c` and before `d`?  I guess I'll fix all of them then.

File: doc/lua_api.txt (hunk 7 above)
Comment (v-rob) [reply doc/lua_api.txt]: Darn it, why do I have to document what should have already been documented in other PRs?

File: src/gui/guiFormSpecMenu.cpp [hunk 9]
```diff
@@ -2162,11 +2162,11 @@ void GUIFormSpecMenu::parseBox(parserData* data, const std::string &element)
 			colors = {tmp_color, tmp_color, tmp_color, tmp_color};
 		} else {
 			colors = style.getColorArray(StyleSpec::COLORS, {0x0, 0x0, 0x0, 0x0});
+			std::array<video::SColor, 4> bordercolors = style.getColorArray(
+				StyleSpec::BORDERCOLORS, {0x0, 0x0, 0x0, 0x0});
+			std::array<s32, 4> borderwidths = style.getIntArray(
+				StyleSpec::BORDERWIDTHS, {0, 0, 0, 0});
```
Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.cpp]: Mistake? `bordercolors`  and `borderwidths` are used outside this scope,

File: doc/lua_api.txt (hunk 6 above)
Comment (SmallJoker) [reply doc/lua_api.txt]: Oh. I only looked at the diff where I assumed it would be sorted. Thanks for sorting them all.

**Comment by SmallJoker** [comment]
Re-tested after the code fix - works.

Q: Is it intended to allow empty values for `colors` or `bordercolors`? (sample from your testing code)
`"style_type[box;colors=;borderwidths=2]"..`
According to the documentation this is not a valid value for this style parameter.

File: src/gui/guiFormSpecMenu.cpp (hunk 9 above)
Comment (v-rob) [reply src/gui/guiFormSpecMenu.cpp]: Yup, mistake.

**Comment by v-rob** [comment]
According to the docs on "Styling Formspecs":

> Setting a property to nothing will reset it to the default value.

https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L2563

So documentation should be unnecessary.  The descriptions for `colors`, `bordercolors`, and `borderwidths` do say their default values (`black`, `black`, and `0`, respectively)

**Comment by v-rob** [comment]
Five month bump.  Just needs one more approval.

**Comment by paramat** [comment]
v-rob, as you are now a core dev, this now has 2 approvals and can be merged.

**Comment by v-rob** [comment]
Actually, I think the code needs some updating first.  It's a bit old.

**Comment by paramat** [comment]
You may know this already, but to merge you would usually use the 'Squash and merge' button shown below.
That will squash all the commits in the PR into one.
That will allow you to edit and clean up the commit message.

That button has an option to 'Rebase and merge'. As far as i remember that is only for when you want to merge without squashing, but i am not completely sure so best discuss before using it.
The dev wiki rules seem a little misleading as they seem to suggest the 'Rebase and merge' option is the primary way to merge, it is not.

**Comment by v-rob** [comment]
Docs had a duplication of some styling properties due to other PRs.  Nothing else needed changing.

**Pull request merged**